### ISSUE TITLE
hipchat module: SSL verification fails with no explanation

### ISSUE THREAD

**Issue by stanhu** [ansible/ansible#6412]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.5.2
##### Environment:

Ubuntu 12.04
##### Summary:

SSL verification fails when attempting to use the hipchat module
##### Steps To Reproduce:

Use the hipchat module and attempt to send a message.
##### Expected Results:

No errors, message gets sent.
##### Actual Results:

```
msg: failed to validate the SSL certificate for api.hipchat.com:443. You can use validate_certs=no, however this is unsafe and not recommended
```

---

Ansible v1.5.2 appears to have enabled SSL server certificate verification. If I attempt to use the HipChat module, it fails due to an unclear error message:

```
msg: failed to validate the SSL certificate for api.hipchat.com:443. You can use validate_certs=no, however this is unsafe and not recommended
```

It appears that Ansible hard-codes the Certificate Authority bundle based on the current platform and distribution. For Ubuntu, the file is hard-coded to /usr/share/ca-certificates/cacert.org/cacert.org.crt.

Running update-ca-certificates on Ubuntu does not fix the problem because this generates a bundle in /etc/ssl/certs/ca-certificates.crt.

I see a few improvements:
1. Display the attempted filename in the error message--this would help clarify what went wrong.
2. Allow configuration of the bundle file in the Ansible configuration (similar to curl's --cacert parameter).
3. curl also allows a --capath directory to search for certificates within this directory. Support this functionality.

**Comment by mpdehaan** [comment]
We're not using curl here but I agree we should look in multiple applications for Ubuntu.   @jimi-c can you take a look at this one to make it look in both locations?

**Comment by mpdehaan** [comment]
Also it does not seem that hipchat module supports validate_certs=no so there's no workaround here.   It seems we would want to add this parameter to any module using the urls.py logic.

Marked as priority -- we should also audit other modules at this time.

**Comment by stanhu** [comment]
Looking in multiple locations seems like a hack; there's no guarantee a user has the ca-certificates module or any of the other certificates installed on the system. I think the right fix is to allow the user to specify a CA file or a CA path, but I don't see an easy way to do this in the current Ansible module structure.  Is there a way to retrieve a config parameter from a module?

By default, it appears that curl uses no CA file but searches the /etc/ssl/certs path:

```
$ curl -v https://api.hipchat.com/v1/room
* About to connect() to api.hipchat.com port 443 (#0)
*   Trying 54.243.83.48... connected
* successfully set certificate verify locations:
*   CAfile: none
  CApath: /etc/ssl/certs
* SSLv3, TLS handshake, Client hello (1):
* SSLv3, TLS handshake, Server hello (2):
* SSLv3, TLS handshake, CERT (11):
* SSLv3, TLS handshake, Server finished (14):
* SSLv3, TLS handshake, Client key exchange (16):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSLv3, TLS change cipher, Client hello (1):
* SSLv3, TLS handshake, Finished (20):
* SSL connection using AES256-SHA
* Server certificate:
*        subject: C=US; ST=California; L=San Francisco; O=Atlassian, Inc.; OU=HipChat; CN=*.hipchat.com
*        start date: 2012-09-26 00:00:00 GMT
*        expire date: 2015-12-20 12:00:00 GMT
*        subjectAltName: api.hipchat.com matched
*        issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=DigiCert High Assurance CA-3
*        SSL certificate verify ok.
```

Unfortunately, the Python ssl package for 2.7 doesn't appear to have any way of passing a capath. It appears that Python 3.x does (http://docs.python.org/3.4/library/ssl.html).

**Comment by mpdehaan** [comment]
So looking in multiple locations is fine as if we don't find any, the verification code will _NOT_ require one, the more greater sin here is validate_certs not being passed.

But you're right, which file would we know it would be in?

We'll discuss this tomorrow, though absolutely without a doubt we should also be exposing validate_cert=no here, and upgrading the message a bit.

**Comment by jimi-c** [comment]
@stanhu I believe the above patch should resolve this issue for you. The function will now look for any .crt or .pem file within common cert paths like `/etc/ssl/certs` and try each in turn until one works (or none do). If you continue to have this problem, please let us know. Thanks!

**Comment by stanhu** [comment]
@jimi-c, thanks for the fix. I still think the path should be configurable because there will be cases where a user will want to use private certificate authorities not listed in the default paths. I'm also not sure if it's a good idea to turn off validation if no certificates can be found. With all the recent security bugs relating to man-in-the-middle attacks, I would err on the side of caution and flag an error, forcing the user to use validate=no if that behavior is desired.

**Comment by stanhu** [comment]
Also, the retry mechanism of the newly-revised function means that connecting to an SSL socket can be very slow. On my Ubuntu system, I count 155 certificates installed in /etc/ssl/certs. I did a quick test: with the new mechanism, it took 42 tries to get the right certificate. This means 42 connection attempts were issued just to initiate a successful SSL connection. What's worse is that certain servers may throttle repeated connection attempts.

I did a quick experiment with curl, and somehow curl managed to find the right certificate in the first open() call:

```
$ strace curl https://api.hipchat.com/v1/room |& grep cert
stat("/etc/ssl/certs/244b5494.0", {st_mode=S_IFREG|0644, st_size=1367, ...}) = 0
open("/etc/ssl/certs/244b5494.0", O_RDONLY) = 4
stat("/etc/ssl/certs/244b5494.1", 0x7fff42e81200) = -1 ENOENT (No such file or directory)
```

I suspect OpenSSL uses the hash value of the certificates to find the right one quickly. This is probably a feature of the --capath argument, and why Python 3.x supports this.

Remember that update-ca-certificates job is to consolidate the different certificates in a single file so that only one connection attempt is needed. You may want to consider putting the bundle as the first file to try. Again, I think this is another reason why configuration options to specify the bundle and cert path in Ansible would be helpful.

**Comment by prune998** [comment]
Having the same issue with flowdoc notification module... the error message seems to come from URLlib...
I don't mind checking the cert of flowdoc... but where to set the variable "validate_certs=no" so ansible can work like it was before 1.5 updates ?

**Comment by jimi-c** [comment]
@prune998 the flowdoc module was updated via another commit, though the above should have fixed the issue for that module as well if you're running off of devel. The `validate_certs=no` is set as a module parameter.

**Comment by mpdehaan** [comment]
We're working on adding that to all modules now on the devel branch and will be doing a 1.5.3 release shortly.

**Comment by prune998** [comment]
thanks... waiting for it... don't want to deploy devel branch on production server... better stop working until then :smile:

**Comment by jimi-c** [comment]
@stanhu here is another possible way of handling it: https://gist.github.com/jimi-c/cc535be8c64a4a8ef7ae, if you want to test that out.

**Comment by stanhu** [comment]
@jimi-c, that certainly would be better than trying all the certificates individually. You're essentially duplicating what Debian's update-ca-certificates does, except not doing it as well. For example, update-ca-certificates handles certificates that are locally installed in /usr/local/share/ca-certificates and also excludes ones that are specified in the config file. Plus, the temporary file is generated for every SSL connection. 

I don't think Ansible should be making these kinds of system-dependent choices. I'm all for reasonable defaults, but at some point the system admin should be able to override them.

**Comment by sivel** [comment]
What if we were to take the route that requests uses.  requests ships a cacert.pem file that is used for cert validation.

That coupled with the configuration option in ansible.cfg and perhaps also looking in /etc/ansible/ could be a decent idea.

https://github.com/kennethreitz/requests/blob/master/requests/cacert.pem
https://github.com/kennethreitz/requests/blob/master/requests/certs.py

**Comment by mpdehaan** [comment]
@sivel, so there are upgrades to this now available on 1.5.3 and devel -- please take a look there.

I don't really want to be in the business of shipping our own cacert.pem, that seems to open up a big can of maintaince worms where nobody is going to be happy with the contents of what we use.

(we can't use requests as that would require an additional dep and complicate bootstrapping)

**Comment by mpdehaan** [comment]
Closing this given updates in 1.5.3, please let us know if you are still seeing problems.
